document-store/src/main/java/org/hypertrace/core/documentstore/postgres/PostgresCollection.java

```diff
@@ -0,0 +1,347 @@
+      preparedStatement.setString(2, jsonString);
+      preparedStatement.setString(3, jsonString);
+      ResultSet resultSet = preparedStatement.executeQuery();
+      
+      if (LOGGER.isDebugEnabled()) {
+        LOGGER.debug("Write result: " + resultSet.toString());
+      }
+      
+      return resultSet.rowUpdated();
+    } catch (SQLException e) {
+      LOGGER.error("SQLException inserting document. key: {} content:{}", key, document, e);
+    }
+    return false;
+  }
+  
+  @Override
+  public boolean updateSubDoc(Key key, String subDocPath, Document subDocument) {
+    return false;
+  }
+  
+  @Override
+  public Iterator<Document> search(Query query) {
+    String filters = null;
+    String space = " ";
+    StringBuilder searchSQLBuilder = new StringBuilder("SELECT * FROM")
+      .append(space).append(collectionName);
+    
+    // If there is a filter in the query, parse it fully.
+    if (query.getFilter() != null) {
+      filters = parseQuery(query.getFilter());
+    }
+    
+    LOGGER.debug(
+      "Sending query to PostgresSQL: {} : {}",
+      collectionName,
+      filters);
+    
+    if (filters != null) {
+      searchSQLBuilder
+        .append(" WHERE ").append(filters);
+    }
+    
+    if (!query.getOrderBys().isEmpty()) {
+      String orderBySQL = parseOrderByQuery(query.getOrderBys());
+      searchSQLBuilder.append(" ORDER BY ").append(orderBySQL);
+    }
+    
+    Integer limit = query.getLimit();
+    if (limit != null && limit >= 0) {
+      searchSQLBuilder.append(" LIMIT ").append(limit);
+    }
+    
+    Integer offset = query.getOffset();
+    if (offset != null && offset >= 0) {
+      searchSQLBuilder.append(" OFFSET ").append(offset);
+    }
+    
+    try {
+      PreparedStatement preparedStatement = client.prepareStatement(searchSQLBuilder.toString());
+      ResultSet resultSet = preparedStatement.executeQuery();
+      return new PostgresResultIterator(resultSet);
+    } catch (SQLException e) {
+      LOGGER.error("SQLException querying documents. query: {}", query, e);
+    }
+    
+    return null;
+  }
+  
+  @VisibleForTesting
+  /**
+   *
+   */
+  String parseQuery(Filter filter) {
+    if (filter.isComposite()) {
+      Filter.Op op = filter.getOp();
+      switch (op) {
+        case OR: {
+          String childList =
+            Arrays.stream(filter.getChildFilters())
+              .map(this::parseQuery)
+              .filter(str -> !str.isEmpty())
+              .collect(Collectors.joining(" OR "));
+          if (!childList.isEmpty()) {
+            return childList;
+          } else {
+            return null;
+          }
+        }
+        case AND: {
+          String childList =
+            Arrays.stream(filter.getChildFilters())
+              .map(this::parseQuery)
+              .filter(str -> !str.isEmpty())
+              .collect(Collectors.joining(" AND "));
+          if (!childList.isEmpty()) {
+            return childList;
```

Comment:
Q: Do we need bracketing here as child filter can also be composite? - `"("+ childList + ")"`
So, case like e.g `(a=2 AND b=3) or (c=4 AND d=5)`, is this fine as `a=2 AND b=3 OR c=4 AND d=5?

Reply:
Yes, added this.